harness/determined/exec/harness.py

```diff
@@ -125,6 +137,117 @@ def main(train_entrypoint: str) -> int:
+        elif distributed_backend.use_deepspeed():
+            distributed = core.DistributedContext.from_deepspeed()
+        elif distributed_backend.use_torch():
+            distributed = core.DistributedContext.from_torch_distributed()
+        elif len(info.container_addrs) > 1 or len(info.slot_ids) > 1:
+            raise ValueError(
+                "In multi-slot tasks, the determined.exec.harness module must not be invoked "
+                "directly.  Instead, it must be wrapped in one of the following launch layers: "
+                "determined.launch.horovod, determined.launch.deepspeed"
+            )
+
+        # Step 4: Let core.init() create the core.Context.
+        with core.init(
+            distributed=distributed,
+            preempt_mode=core.PreemptMode.ChiefOnly,
+            tensorboard_mode=core.TensorboardMode.MANUAL,
+        ) as core_context:
+            trial_context = pytorch.PyTorchTrialContext(
+                core_context=core_context,
+                trial_seed=info.trial.trial_seed,
+                hparams=info.trial.hparams,
+                slots_per_trial=info.trial._config["resources"]["slots_per_trial"],
+                num_gpus=len(info.gpu_uuids),
+                exp_conf=info.trial._config,
+                aggregation_frequency=info.trial._config["optimizations"]["aggregation_frequency"],
+                fp16_compression=info.trial._config["optimizations"]["gradient_compression"],
+                average_aggregated_gradients=cast(
+                    bool, info.trial._config["optimizations"]["average_aggregated_gradients"]
+                ),
+                steps_completed=info.trial._steps_completed,
+                managed_training=True,
+            )
+
+            # Step 4: Instantiate the user's Trial.
+            trial_inst = trial_class(trial_context)
+
+            # Step 5: Create a TrialController and execute training
+            logging.info(
+                f"Creating {pytorch.PyTorchTrialController.__name__} with {trial_class.__name__}."
+            )
+
+            profiling_enabled = cast(bool, info.trial._config["profiling"]["enabled"])
+            det_profiler = ProfilerAgent(
+                trial_id=str(info.trial.trial_id),
+                agent_id=info.agent_id,
+                master_url=info.master_url,
+                profiling_is_enabled=profiling_enabled,
+                global_rank=core_context.distributed.rank,
+                local_rank=core_context.distributed.local_rank,
+                begin_on_batch=profiling_enabled
+                and cast(int, info.trial._config["profiling"]["begin_on_batch"])
+                or 0,
+                end_after_batch=profiling_enabled
+                and cast(int, info.trial._config["profiling"]["end_after_batch"])
+                or 0,
+                sync_timings=cast(bool, info.trial._config["profiling"]["sync_timings"]),
+            )
+
+            controller = pytorch.PyTorchTrialController(
```

Comment:
I may have asked you this already... but why doesn't `harness.py` use `det.pytorch.init()` anywhere?

Reply:
harness now uses trainer.